test/test_thread_safety.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+require_relative 'helper'
+
+Thread.abort_on_exception = true
+
+class ThreadSafetyTest < Minitest::Test
+  include ARHPTestSetup
+  def setup
+    if RAILS_6_1_WITH_NON_LEGACY_CONNECTION_HANDLING
+      Phenix.rise! config_path: 'test/three_tier_database.yml'
+    else
+      Phenix.rise!
+    end
+    arhp_create_models
+
+    Pool1DbA.create!(val: 'test_pool_1_DB_A_value')
+    Pool1DbB.create!(val: 'test_pool_1_DB_B_value')
+    Pool2DbD.create!(val: 'test_pool_2_DB_D_value')
+  end
+
+  def teardown
+    ActiveRecord::Base.connection.disconnect!
+    ActiveRecordHostPool::PoolProxy.class_variable_set(:@@_connection_pools, {})
+    Phenix.burn!
+  end
+
+  def test_main_and_spawned_thread_switch_db_when_querying_same_host
+    assert_query_host_1_db_A
+
+    thread = Thread.new do
+      assert_query_host_1_db_B
+
+      Thread.current[:done] = true
+      sleep
+
+      checkin_connection
+    end
+
+    sleep 0.01 until thread[:done]
+
+    assert_query_host_1_db_A
+
+    thread.wakeup
+    thread.join
+  end
+
+  def test_main_and_spawned_thread_can_query_different_hosts
+    assert_query_host_1_db_A
+
+    thread = Thread.new do
+      assert_query_host_2_db_D
+
+      Thread.current[:done] = true
+      sleep
+
+      checkin_connection
+    end
+
+    sleep 0.01 until thread[:done]
+
+    assert_query_host_1_db_A
+
+    thread.wakeup
+    thread.join
+  end
+
+  def test_threads_can_query_in_parallel
+    long_sleep = 0.5
+    short_sleep = 0.1
+
+    even_threads_do_this = [
+      { method: :assert_query_host_1_db_A, db_sleep_time: long_sleep },
+      { method: :assert_query_host_1_db_B, db_sleep_time: short_sleep }
+    ]
+    odd_threads_do_this = [
+      { method: :assert_query_host_1_db_B, db_sleep_time: short_sleep },
+      { method: :assert_query_host_1_db_A, db_sleep_time: long_sleep }
+    ]
+
+    threads = 4.times.map do |n|
+      Thread.new do
+        Pool1DbA.connection
+        Thread.current[:ready] = true
+        sleep
+
+        Thread.current.name = "Test thread #{n}"
+
+        what_to_do = n.even? ? even_threads_do_this : odd_threads_do_this
```

Comment:
Should these just be lambdas?

This way is good for experimenting with different permutations but if it's locked in like this then lambdas might be easier to understand.

Reply:
I think you're right that lambdas would make it more intuitive! Maybe let's do that and then I can merge this in?